### PR TITLE
Tx verifiers simulation test

### PR DESCRIPTION
**Pull request by azarovh**
- Introduced `RandomizedTransactionVerificationStrategy` that would derive, connect/disconnect, flush in random order.

- Added `tx_verification_simulation` test that would create and process random number of blocks and transactions. Main functionality for this test is in `create_multiple_utxo_data` function that can create random number of different outputs based on the single input provided. 

- Amount of atoms in the genesis output was increased for test network. It was only enough to issue 10 tokens and was quickly exhausted in simulation.

### CODE REVIEW

File: chainstate/storage/src/internal/mod.rs [hunk 1]
```diff
@@ -76,6 +76,18 @@ impl<B: storage::Backend> Store<B> {
     pub fn dump_raw(&self) -> crate::Result<storage::raw::StorageContents<Schema>> {
         self.0.dump_raw().map_err(crate::Error::from)
     }
+
+    /// Collect and return all utxos from the storage
+    #[allow(clippy::let_and_return)]
```
Comment (stanislav-tkach) [inline chainstate/storage/src/internal/mod.rs:81]: You can submit an issue about this false positive if you want. 🙃

**Review by stanislav-tkach — APPROVED** [review]

File: chainstate/test-framework/src/test_block_info.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+// limitations under the License.
+
+use crate::TestChainstate;
+use chainstate::chainstate_interface::ChainstateInterface;
+use common::{
+    chain::{gen_block::GenBlockId, Block, GenBlock, Genesis, OutPointSourceId, TxOutput},
+    primitives::{Id, Idable},
+};
+
+// TODO: Replace by a proper UTXO set abstraction
```
Comment (TheQuantumPhysicist) [inline chainstate/test-framework/src/test_block_info.rs:23]: This is still not achievable with the current changes?

File: chainstate/test-framework/src/tx_verification_strategy/randomized_strategy.rs [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            )
+            .log_err()?;
+        debug_assert!(reward_fees.is_none());
+
+        let mut total_fee = Amount::ZERO;
+        let mut tx_num = 0usize;
+        while tx_num < block.transactions().len() {
+            let switch = self.rng.borrow_mut().gen_range(0..5);
+            if switch == 0 {
+                // derive a new cache in 20% cases
```
[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply chainstate/test-framework/src/tx_verification_strategy/randomized_strategy.rs:150]: Once you do that, please run the tests like 100 times to ensure that we don't have uncovered parameter spaces with that setting.

**Review by TheQuantumPhysicist — APPROVED** [review]
Looks good. Please mind the few comments.

File: chainstate/test-framework/src/tx_verification_strategy/randomized_strategy.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (azarovh) [reply chainstate/test-framework/src/tx_verification_strategy/randomized_strategy.rs:150]: Do you suggest doing something like this?
``` rust 
let p = self.rng.borrow_mut().gen_range(0.0_f64..1.0_f64);
let switch = self.rng.borrow_mut().gen_bool(p);
```

In other places as well?

File: chainstate/storage/src/internal/mod.rs (hunk 1 above)
Comment (azarovh) [reply chainstate/storage/src/internal/mod.rs:81]: It's already there:
https://github.com/rust-lang/rust-clippy/issues/8114

File: chainstate/test-framework/src/tx_verification_strategy/randomized_strategy.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheQuantumPhysicist) [reply chainstate/test-framework/src/tx_verification_strategy/randomized_strategy.rs:150]: Yes. This looks fine. 

Do you have to do it in every spot? I'll leave it up to your discretion. But values up your sleeve, like 20%, seem like a bad idea. No need to fix something without a good reason.

[the remark this answers is not in the document]
Comment (iljakuklic) [reply chainstate/test-framework/src/tx_verification_strategy/randomized_strategy.rs:150]: > Do you suggest doing something like this?
> 
> ```rust
> let p = self.rng.borrow_mut().gen_range(0.0_f64..1.0_f64);
> let switch = self.rng.borrow_mut().gen_bool(p);
> ```
> 
> In other places as well?

I think this is equivalent to just generating a bool that is `true` with 50% probability (unless the value of `p` is used in multiple places)

File: chainstate/test-framework/src/test_block_info.rs (hunk 2 above)
Comment (stanislav-tkach) [reply chainstate/test-framework/src/test_block_info.rs:23]: We have an issue for this: https://github.com/mintlayer/mintlayer-core/issues/312
Perhaps @iljakuklic can provide more details if someone wants to improve this part.

**Pull request merged**